themes/default/content/blog/benchmarking-python-performance/index.md

```diff
@@ -0,0 +1,222 @@
+
+<!--more-->
+
+We recently took a hard look at the performance of Python programs when we
+realized they weren't performing up to our expectations. We uncovered a major
+bug limiting Python performance, and we ran a number of rigorous experiments
+to evaluate just how performant Pulumi Python programs are after the bug had
+been repaired. The results indicate Pulumi Python programs are significantly
+faster than they were, and now Pulumi Python has reached performance parity
+with Pulumi Node!
+
+## The Bug
+
+When you execute a Pulumi program, Pulumi internally builds a dependency graph
+between the resources in your program. In every Pulumi program, some resources
+have all their input arguments available at the time of their construction.
+In contrast, other resources may depend on `Outputs` from other resources.
+
+For example, consider a sample program where we expose a compute instance
+publicly using a floating IP:
+
+```javascript
+import * as pulumi from "@pulumi/pulumi";
+import * as digitalocean from "@pulumi/digitalocean";
+
+const myDroplet = new digitalocean.Droplet("MyDroplet", {
+    size: "s-1vcpu-1gb",
+    image: "ubuntu-18-04-x64",
+    region: "sgp1",
+    ipv6: true,
+    privateNetworking: true,
+});
+const myFloatingIp = new digitalocean.FloatingIp("MyFloatingIp", {
+    dropletId: foobarDroplet.id,
+    region: foobarDroplet.region,
+});
+```
+
+Because the `FloatingIp` takes an input an `Output` typed value from
+`myDroplet`, we can't create the `FloatingIP` until after the `Droplet`
+is created. We have to create the `Droplet` first to compute the `Droplet`'s ID,
+which we can pass to the `FloatingIp`. This idea extends inductively for
+arbitrary programs – before any resource can be run, we must resolve the
+`Outputs` of all of its arguments. To do this, Pulumi builds a dependency graph
+between all resources in your program. Then, it walks the graph topologically
+to schedule provisioning operations. 
+
+Provisioning operations that are not dependent on each other can be executed
+in parallel, and Pulumi defaults to unbounded parallelism, but users can
+ratchet this down if they so desire. Consider this embarrassingly parallel
+Python program:
+
+```python
+import pulumi
+import pulumi_aws as aws
+
+# SQS
+for i in range(100):
+	name = f'pulumi-{str(i).rjust(3, "0")}'
+	aws.sqs.Queue(name)
+
+# SNS
+for i in range(100):
+	name = f'pulumi-{str(i).rjust(3, "0")}'
+	aws.sns.Topic(name)
+```
+
+In this program, we can create 200 resources in parallel because none of them
+take inputs from other resources. This program should be entirely
+network-bound because Pulumi can issue all 200 API calls in parallel and wait 
+for AWS to provision the instances.
+[We discovered](https://github.com/pulumi/pulumi/issues/11116), however, 
+that it did not! Strangely, API calls were issued in an initial batch of 20; 
+as one completed, another would start.
+
+The culprit was the Python default future executor,
```

Comment:
This section simultaneously gets too far in the weeds _and_ buries the lede. This was the big fix that gave us 300% perf improvement, but at the same time it's not the focus of the blog post, the experiments are. Not quite sure how I want to modulate the tone here.